Filter Sections input field added

MrAardbij opened the pull request:
Hello,
I added an extra field in the dialog so you can filter the sections.
Useful when you have a lot of sections.
Would love to hear feedback.

MrAardbij committed:
Filter Sections input field added

VinhPhmCng commented on addons/gdscript_sections/dialog.gd:
```diff
@@ -82,16 +87,20 @@ func _on_DisableAccept_pressed() -> void:
-	add_section.clear()
-	return
-
-
```
Why is this feature being removed?

VinhPhmCng commented on addons/gdscript_sections/dialog.gd:
```diff
@@ -47,6 +49,9 @@ func set_font_size(to: int) -> void:
 
+func _handle_close_event() -> void:
+	filter_section.clear()
+	return
```
Clearing text on closing is not in accordance with the Godot engine (e.g. text fields in Project Settings are left unchanged even if the user closes the popup) and therefore unnecessary.

VinhPhmCng commented on addons/gdscript_sections/gdscript_sections.gd:
```diff
@@ -409,8 +411,30 @@ func _on_Dialog_section_added(text: String) -> void:
 	)
 
 	_update_ui_and_display()
+	_section_filter_ui_reset()
```
Clearing the filter field when adding a new section is also unnecessary. 

You can check Godot's Input Map to see how its text fields handle events.

VinhPhmCng commented on addons/gdscript_sections/gdscript_sections.gd:
```diff
@@ -409,8 +411,30 @@ func _on_Dialog_section_added(text: String) -> void:
 
+## filters the current ui
+func _on_Dialog_section_filter(text: String) -> void:
+	_update_OverlayDisplay()
```
This line doesn't do anything.

VinhPhmCng commented on addons/gdscript_sections/gdscript_sections.gd:
```diff
@@ -409,8 +411,30 @@ func _on_Dialog_section_added(text: String) -> void:
+	var root := tree.create_item()
+
+	for section in sections:
+		if section.text.to_lower().begins_with(text.to_lower()):
```
This filter logic isn't satisfactory.

We should replicate the logic used by Godot engine (e.g. the field `Filter by name...` in Input Map). 
This can be easily achieved with RegEx. I can work on this when I have the time.

VinhPhmCng commented:
Thanks for the PR, bro. The filter feature is certainly useful.

I've made some suggestions here and there. The only thing to be improved is the filter logic, which I can work on soon enough.

MrAardbij committed:
restored on-enter submit

MrAardbij committed:
redone filter so it doesnt get cleared on close

MrAardbij commented:
Hello,
I dont know why i removed the text_submitted event, re-added. Apologies.

There is added value i think on having the option of clearing the filter on window close because it changes a lot.
Same story with clearing the filter field when adding a new section as the new value disappears immediately.
Maybe make it a setting?

I have redone the filter function so it calls _update_ui_and_display() and handles the filter logic with _check_section_on_filter(section: Section) -> bool
The filter now calls String.contains()

VinhPhmCng committed:
Filter logic re-implemented

Filter logic re-implemented using RegEx

Code cleaned

VinhPhmCng commented:
> Hello, I dont know why i removed the text_submitted event, re-added. Apologies.

No worries.

> There is added value i think on having the option of clearing the filter on window close because it changes a lot. Same story with clearing the filter field when adding a new section as the new value disappears immediately. Maybe make it a setting?

In my opinion, it's redundant because not clearing the fields is more convenient in most cases.
For now, I've enabled `Clear button` for both `LineEdit`s.

> I have redone the filter function so it calls _update_ui_and_display() and handles the filter logic with _check_section_on_filter(section: Section) -> bool The filter now calls String.contains()

I've re-implemented the filter logic using RegEx. It should behave identically to that of Godot now.

Great! Good job @MrAardbij. 
Check the commit out. If nothing's wrong, I'll merge.

VinhPhmCng committed:
Added checking for escaped characters in filter

Added checking for escaped characters so that they are not mistaken as part of the expression

MrAardbij commented:
Can you check the following?
Dont know if this is a godot problem or regex thing.

if you have a number in the section name and type it in the filter box you get following error:
  modules/regex/regex.cpp:198 - 17: reference to non-existent subpattern
  modules/regex/regex.cpp:205 - Condition "!is_valid()" is true. Returning: nullptr

and it gives no results

![image](https://github.com/VinhPhmCng/gdscript-sections/assets/43466799/b37daa60-fdf2-4f73-95b7-40de74091e5b)

MrAardbij commented:
the regex:
s.*?g.*?l.*?s.*?\1.*?

VinhPhmCng committed:
Improved checking for escaped characters in filter

VinhPhmCng commented:
It should be good now.

Anything else?
